docs/unification.md

```diff
@@ -0,0 +1,344 @@
+improper binding x = y, we substitute y for x everywhere in the solution and
+add `x = y` to the list of improper bindings. When substituting, if we
+substitute in a manner such that we are left with an equality that is between
+two non-variables, we remove it from the solution and add it to the list of
+free equations to decompose.
+
+Note that there are two ways to consider an equation `x = y`, since equalities
+are commutative. We could replace `x` with `y` or `y` with `x`. In practice, we
+always replace `x` with `y` because we orient the bindings generated by
+variable abstraction such that `x` would always correspond to the fresh
+variable generated by variable abstraction, which is always considered an
+acceptable orientation.
+
+Set
+---
+
+Set unification proceeds identical to map unification except that we treat each
+set element as a map element from its value to ().
+
+List
+----
+
+If both terms are function symbols of list sort, we conside 5 cases:
+
+1. If one list is `_List_(l1, Var1)` and the other list is `_List_(l2, Var2)`, where l1 and l2 have no opaque terms, and len(l1) <= len(l2), for i = 1 to len(l1), decompose with the ith element of l1 and l2. Then, remove len(l1) elements from the start of l2 and decompose with Var1 and `_List_(l2, Var2)`
+2. If one list is `_List_(Var1, l1)` and the other list is `_List_(Var2, l2)`, where l1 and l2 have no opaque terms, and len(l1) <= len(l2), for i = 1 to len(l1), decompose with the ith element of l1 and the len(l2) - len(l1) + ith element of l2. Then, remove len(l1) elements from the end of l2 and decompose with Var1 and `_List_(Var2, l2)`
+3. If both lists have no opaque terms, and they have the same length, unify each element in sequence. If they have different lengths, fail.
+4. If one list l1 has no opaque terms and the other is `_List_(l2, Var)`, and len(l2) <= len(l1), then for i = 1 to len(
+l2), decompose with the ith element of l1 and l2. Then, remove len(l2) elements from the start of l1 and decompose with Var and l1. If len(l2) > len(l1), fail.
+4. If one list l1 has no opaque terms and the other is `_List_(Var, l2)`, and len(l2) <= len(l1), then for i = 1 to len(l2), decompose with the ith element of l2 and the len(l1) - len(l2) + ith element of l1. Then, remove len(l2) elements from the end of l1 and decompose with Var and l1. If len(l2) > len(l1), fail.
+5. Otherwise, add the constraint that the two lists are equal.
+
+#if #then #else #fi
+-------------------
+
+This is a case that properly speaking, probably ought to be expressed as a
+simplification rule, but historically, it was made to be part of the
+unification algorithm and writing it as a rule in K is impossible due to the
+current restrictions on parametric productions in K. It is quite
+straightforward, however:
+
+If one term is `#if B #then T #else F #fi` for patterns B, T, F, and the other
+term is `P`, then we fork with two cases:
+
+1. Decompose with T and P and constrain with `\ceil(\and(B, true))`
```

Comment:
Does this work if pattern `F` may not be defined? I.e., `ceil(and(#if B #then T #else Bottom #fi, P)) == ceil(and(Bottom, P)) == Bottom`, while, with this decomposition, the end result seems to be `ceil(and(T, P)) and \ceil(\and(B, true))`

Reply:
I'm afraid once again, I don't fully follow the example you mentioned, but this case is merely faithfully rendering one of the cases already implemented in the unification algorithm. If it's unsound, then you are welcome to raise an issue with the Haskell backend team and it can be removed, but doing so seems outside the scope of this PR to me.